clippy_lints/src/methods/inspect_then_for_each.rs

```diff
@@ -1,20 +1,46 @@
+use if_chain::if_chain;
 use rustc_hir as hir;
 use rustc_lint::LateContext;
+use rustc_span::source_map::Span;
 
-use crate::utils::{match_trait_method, paths, span_lint_and_help};
+use crate::utils::{match_trait_method, paths, span_lint_and_help, SpanlessEq};
 
 use super::INSPECT_THEN_FOR_EACH;
 
 /// lint use of `inspect().for_each()` for `Iterators`
 pub(super) fn lint<'tcx>(
     cx: &LateContext<'tcx>,
     expr: &'tcx hir::Expr<'_>,
-    _inspect_args: &'tcx [hir::Expr<'_>],
-    _for_each_args: &'tcx [hir::Expr<'_>],
+    inspect_args: &'tcx [hir::Expr<'_>],
+    for_each_args: &'tcx [hir::Expr<'_>],
+    inspect_span: Span,
 ) {
     if match_trait_method(cx, expr, &paths::ITERATOR) {
-        let msg = "called `inspect(..).for_each(..)` on an `Iterator`";
-        let hint = "this is more succinctly expressed by calling ` for_each(..)` instead";
-        span_lint_and_help(cx, INSPECT_THEN_FOR_EACH, expr.span, msg, None, hint);
+        let inspect_closure = &inspect_args[1].kind;
+        let for_each_closure = &for_each_args[1].kind;
+
+        if_chain! {
+            if let hir::ExprKind::Closure(_, _, inspect_closure_body_id, _, _) = inspect_closure;
+            if let hir::ExprKind::Closure(_, _, for_each_closure_body_id, _, _) = for_each_closure;
+
+            let inspect_closure_pat = cx.tcx.hir().body(*inspect_closure_body_id).params[0].pat;
+            let for_each_closure_pat = cx.tcx.hir().body(*for_each_closure_body_id).params[0].pat;
+
+            if SpanlessEq::new(cx).eq_pat(inspect_closure_pat, for_each_closure_pat);
```

Comment:
This is probably not necessary, but since we're already in an `if_chain!` we might as well be safe here.

```suggestion
            if let Some(inspect_closure_param) = cx.tcx.hir().body(*inspect_closure_body_id).params.get(0);
            if let Some(for_each_closure_param) = cx.tcx.hir().body(*for_each_closure_body_id).params.get(0);

            if SpanlessEq::new(cx).eq_pat(inspect_closure_param.pat, for_each_closure_param.pat);
```